### PR TITLE
feat: add snowflake grant privileges to account role

### PR DESCRIPTION
**Pull request by sfc-gh-jcieslak**
"Refactor" of the snowflake_grant_privileges_to_role:
- rename to snowflake_grant_privileges_to_account_role
- new identifier
- documentation for the new resource
- always_apply flag
- validation change in the new resource (according to the spec)

## Test Plan

* [x] unit tests for the new identifier

* [x] acceptance tests for the new snowflake_grant_privileges_to_account_role

## TODO

* Write a migration guide for the new resource as well as guidance on how to migrate from old grant resources and how to specify identifiers in the new ones. Why new resource (name conv + new internal identifier).
* Handle IMPORTED PRIVILEGES and on_all PIPES cases

### CODE REVIEW

File: pkg/resources/grant_privileges_to_account_role_acceptance_test.go [hunk 1]
```diff
@@ -0,0 +1,909 @@
...
+				ConfigVariables:   configVariables,
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAcc_GrantPrivilegesToAccountRole_UpdatePrivileges(t *testing.T) {
```
Comment (sfc-gh-swinkler) [inline pkg/resources/grant_privileges_to_account_role_acceptance_test.go:550]: is this a sufficient amount of update tests? there are many possible combinations, so does this cover all the edge cases?

File: pkg/resources/grant_privileges_to_account_role_acceptance_test.go [hunk 2]
```diff
@@ -0,0 +1,909 @@
...
+					resource.TestCheckResourceAttr(resourceName, "privileges.1", string(sdk.AccountObjectPrivilegeMonitor)),
+					resource.TestCheckResourceAttr(resourceName, "id", fmt.Sprintf("%s|false|false|MODIFY,MONITOR|OnAccountObject|DATABASE|%s", roleName, databaseName)),
+				),
+			},
+		},
+	})
+}
+
+func TestAcc_GrantPrivilegesToAccountRole_UpdatePrivileges_SnowflakeChecked(t *testing.T) {
+	name := "test_account_role_name"
```
Comment (sfc-gh-swinkler) [inline pkg/resources/grant_privileges_to_account_role_acceptance_test.go:640]: should this be a randomized id?

Comment (sfc-gh-asawicki) [reply pkg/resources/grant_privileges_to_account_role_acceptance_test.go:640]: I agree, even just a few random chars at the end

File: pkg/resources/grant_privileges_to_account_role_acceptance_test.go (hunk 1 above)
Comment (sfc-gh-asawicki) [reply pkg/resources/grant_privileges_to_account_role_acceptance_test.go:550]: I don't think we can or even should cover all the cases now with the acceptance tests: they run for a considerable amount of time, it would be testing provider (resource) logic with SF behavior, and it's a bad practice to check all the edge cases in acceptance tests (we should have a test pyramid and not the inverted pyramid). I think that instead we should:
- treat acceptance tests as the smoke tests currently
- introduce more tests (but unit ones) for the provider but in an isolation from SDK (so that we check only how provider behaves gives config + state.

We have a task where we will revisit such discussions. For now, let's not overflow the codebase with all the possible edge cases in acceptance tests.

File: pkg/resources/grant_privileges_to_account_role.go [hunk 3]
```diff
@@ -0,0 +1,1103 @@
...
+				Severity: diag.Error,
+				Summary:  "Failed to parse internal identifier",
+				Detail:   fmt.Sprintf("Id: %s\nError: %s", d.Id(), err.Error()),
+			},
+		}
+	}
+	logging.DebugLogger.Printf("[DEBUG] Parsed identifier: %s", id.String())
+
+	if id.AlwaysApply {
+		triggerId, err := uuid.GenerateUUID()
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/resources/grant_privileges_to_account_role.go:684]: @sfc-gh-jcieslak I suggested it for the previous change and I agree with Scott, that this should be documented in code as a comment because the motivation is unclear.

File: docs/resources/grant_privileges_to_account_role.md [hunk 4]
```diff
@@ -0,0 +1,362 @@
+---
+# generated by https://github.com/hashicorp/terraform-plugin-docs
+page_title: "snowflake_grant_privileges_to_account_role Resource - terraform-provider-snowflake"
+subcategory: ""
+description: |-
+  
+---
+
+
+!> **Warning** Be careful when using `always_apply` field. It will always produce a plan (even when no changes were made) and can be harmful in some setups. For more details why we decided to introduce it to go our document explaining those design decisions (coming soon).
```
Comment (sfc-gh-asawicki) [inline docs/resources/grant_privileges_to_account_role.md:10]: I think it would be nice to add here the same communication I have suggested for the previous resource: https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/2306#discussion_r1441653749 (and I also missed that it was not added there, so please cover both resources)

File: pkg/resources/grant_privileges_to_role.go [hunk 5]
```diff
@@ -129,7 +129,7 @@ var grantPrivilegesToRoleSchema = map[string]*schema.Schema{
 					RequiredWith:     []string{"on_schema_object.0.object_name"},
 					ConflictsWith:    []string{"on_schema_object.0.all", "on_schema_object.0.future"},
 					ForceNew:         true,
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_privileges_to_role.go:131]: should we deprecate this whole object and mention the one with new name?

File: pkg/sdk/grants_validations.go [hunk 6]
```diff
@@ -70,8 +70,8 @@ func (v *AccountRoleGrantOn) validate() error {
 }
 
 func (v *GrantOnAccountObject) validate() error {
-	if !exactlyOneValueSet(v.User, v.ResourceMonitor, v.Warehouse, v.Database, v.Integration, v.FailoverGroup, v.ReplicationGroup, v.ExternalVolume) {
-		return errExactlyOneOf("GrantOnAccountObject", "User", "ResourceMonitor", "Warehouse", "Database", "Integration", "FailoverGroup", "ReplicationGroup", "ExternalVolume")
+	if !exactlyOneValueSet(v.User, v.ResourceMonitor, v.Warehouse, v.Database, v.Integration, v.Connection, v.FailoverGroup, v.ReplicationGroup, v.ExternalVolume) {
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/grants_validations.go:73]: no test changed after this alteration? 🤔

File: pkg/resources/grant_privileges_identifier_commons.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+	return strings.Join(parts, helpers.IDDelimiter)
+}
+
+type OnSchemaObjectGrantData struct {
+	Kind          OnSchemaObjectGrantKind
+	Object        *sdk.Object
+	OnAllOrFuture *BulkOperationGrantData
+}
+
+func (d *OnSchemaObjectGrantData) String() string {
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/resources/grant_privileges_identifier_commons.go:51]: I am also a bit confused now: we have grant_privileges_identifier_commons, grant_privileges_to_database_role_identifier. and grant_privileges_to_account_role_identifier. As far I can see, we have test files only for the last two. What about the commons? Are the tests included in both of other files or are they missing as Scott suggested?

File: pkg/resources/grant_privileges_to_account_role_identifier_test.go [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+				Privileges:      []string{"CREATE SCHEMA", "USAGE", "MONITOR"},
+				Kind:            OnAccountObjectAccountRoleGrantKind,
+				Data: &OnAccountObjectGrantData{
+					ObjectType: sdk.ObjectTypeDatabase,
+					ObjectName: sdk.NewAccountObjectIdentifier("database-name"),
+				},
+			},
+		},
+		{
+			Name:       "grant database role on schema with schema name",
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_privileges_to_account_role_identifier_test.go:67]: database role -> account role; also in some other tests in this file

File: pkg/resources/grant_privileges_to_account_role_identifier_test.go [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+		},
+		{
+			Name:       "validation: account database role empty type",
+			Identifier: `"database-name"."database-role"|false|false|ALL PRIVILEGES||"on-database-name"`,
+			Error:      "invalid AccountRoleGrantKind: ",
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.Name, func(t *testing.T) {
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_privileges_to_account_role_identifier_test.go:302]: shouldn't we `tt := tt` before running these? I guess it's fine without parallelization but it would hurt us if someone decides to add `t.Parallel()` here. I think this applies also to database role tests

File: pkg/resources/grant_privileges_to_account_role_acceptance_test.go [hunk 10]
```diff
@@ -0,0 +1,909 @@
+package resources_test
+
+import (
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_privileges_to_account_role_acceptance_test.go:3]: did you check the old resource with the new tests (ofc beside the new behavior)?

File: pkg/resources/grant_privileges_to_account_role_acceptance_test.go [hunk 11]
```diff
@@ -0,0 +1,909 @@
...
+		}
+		if len(grantedPrivileges) > 0 {
+			return fmt.Errorf("account role (%s) is still granted, granted privileges %v", id.FullyQualifiedName(), grantedPrivileges)
+		}
+	}
+	return nil
+}
+
+// queriedAccountRolePrivilegesEqualTo will check if all the privileges specified in the argument are granted in Snowflake.
+func queriedAccountRolePrivilegesEqualTo(roleName sdk.AccountObjectIdentifier, privileges ...string) func(s *terraform.State) error {
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_privileges_to_account_role_acceptance_test.go:858]: could it be merged with `queriedPrivilegesEqualTo` (the one from database role tests), by just providing the different show provider? the logis besides the show invocation is the same. Providing only `&sdk.ShowGrantsTo` differently should also work.

File: pkg/resources/grant_privileges_to_account_role_acceptance_test.go [hunk 12]
```diff
@@ -0,0 +1,909 @@
...
+			}
+		}
+
+		return nil
+	}
+}
+
+// queriedAccountRolePrivilegesContainAtLeast will check if all the privileges specified in the argument are granted in Snowflake.
+// Any additional grants will be ignored.
+func queriedAccountRolePrivilegesContainAtLeast(roleName sdk.AccountObjectIdentifier, privileges ...string) func(s *terraform.State) error {
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_privileges_to_account_role_acceptance_test.go:883]: same as above

File: pkg/resources/grant_privileges_to_account_role_acceptance_test.go [hunk 13]
```diff
@@ -0,0 +1,909 @@
+package resources_test
+
+import (
+	"context"
+	"database/sql"
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_privileges_to_account_role_acceptance_test.go:5]: the tests look similar to database role ones, I have to admit I skimmed through them. They look fine from the names I did not validate in detail the checks

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: docs/resources/grant_privileges_to_database_role.md [hunk 14]
```diff
@@ -6,6 +6,8 @@ description: |-
   
 ---
 
+!> **Warning** This is a preview resource. It's not production-ready (yet), and some of the edge cases may not be covered. In case of any errors, please file an issue in our GitHub repository.
```
Comment (sfc-gh-asawicki) [inline docs/resources/grant_privileges_to_database_role.md:9]: Why these two warnings are different (this one and the one from account role)? This one seems a bit too hard. I would like people to use it on production to give feedback. I think we can weaken this one.

I am also wandering if these should be warnings or notes (yellow vs blue text boxes). WDYT?

File: pkg/resources/grant_privileges_to_account_role.go [hunk 15]
```diff
@@ -0,0 +1,1109 @@
...
+			"on_schema",
+			"on_schema_object",
+		},
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"object_type": {
+					Type:        schema.TypeString,
+					Required:    true,
+					ForceNew:    true,
+					Description: "The object type of the account object on which privileges will be granted. Valid values are: USER | RESOURCE MONITOR | WAREHOUSE | DATABASE | INTEGRATION | CONNECTION | FAILOVER GROUP | REPLICATION GROUP | EXTERNAL VOLUME",
```
Comment (sfc-gh-asawicki) [inline pkg/resources/grant_privileges_to_account_role.go:98]: CONNECTION -> COMPUTE POOL

Comment (sfc-gh-asawicki) [reply pkg/resources/grant_privileges_to_account_role.go:98]: And maybe in some other places, please check it

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: pkg/resources/grant_privileges_to_account_role.go (hunk 15 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/grant_privileges_to_account_role.go:98]: Done

**Review by sfc-gh-asawicki — APPROVED** [review]

**Review by sfc-gh-swinkler — DISMISSED** [review]

File: docs/resources/grant_privileges_to_account_role.md [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+}
+
+##################################
+### on account privileges
+##################################
+
+# list of privileges
+resource "snowflake_grant_privileges_to_account_role" "example" {
+  privileges = ["CREATE DATABASE", "CREATE USER"]
+  role_name  = snowflake_role.db_role.name
```
Comment (sfc-gh-asawicki) [inline docs/resources/grant_privileges_to_account_role.md:36]: We changed role_name to account_role_name. Please change the examples.

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: docs/resources/grant_privileges_to_account_role.md (hunk 16 above)
Comment (sfc-gh-jcieslak) [reply docs/resources/grant_privileges_to_account_role.md]: DOne

**Review by sfc-gh-asawicki — APPROVED** [review]
